keeper/src/main/scala/zio/keeper/consensus/Consensus.scala

```diff
@@ -0,0 +1,172 @@
+package zio.keeper.consensus
+
+import java.util.UUID
+
+import zio._
+import zio.clock.Clock
+import zio.keeper.membership.{ ByteCodec, Membership, MembershipEvent, NodeId, TaggedCodec }
+import zio.keeper.{ Error, Message, membership }
+import zio.stream.{ Stream, ZSink }
+import zio.keeper.protocol.InternalProtocol._
+import upickle.default._
+import zio.duration.Duration
+import zio.logging.Logging
+
+object Consensus {
+
+  trait Service {
+    def getLeader: UIO[NodeId]
+    def onLeaderChange: Stream[Error, NodeId]
+    def getView: UIO[List[NodeId]]
+    def selfNode: UIO[NodeId]
+    def receive: Stream[Error, Message]
+    def broadcast(data: Chunk[Byte]): IO[Error, Unit]
+    def send(data: Chunk[Byte], to: NodeId): IO[Error, Unit]
+  }
+
+}
+
+/**
+ *  Consensus live where leader is the oldest member in cluster
+ */
+object Coordinator {
+
+  sealed private trait ConsensusMsg
+  final private case object ConsensusViewRequest                     extends ConsensusMsg
+  final private case class ConsensusViewResponse(view: List[NodeId]) extends ConsensusMsg
+  final private case class ConsensusUserMsg(payload: Chunk[Byte])    extends ConsensusMsg
+  private val viewRequestCodec  = ByteCodec.fromReadWriter(macroRW[ConsensusViewRequest.type])
+  private val viewResponseCodec = ByteCodec.fromReadWriter(macroRW[ConsensusViewResponse])
+
+  private val userMsgCodec = ByteCodec.fromReadWriter(
+    implicitly[ReadWriter[Array[Byte]]]
+      .bimap[ConsensusUserMsg](_.payload.toArray, bA => ConsensusUserMsg(Chunk.fromArray(bA)))
+  )
+
+  implicit private val consensusCodec: TaggedCodec[ConsensusMsg] =
+    TaggedCodec.instance[ConsensusMsg](
+      {
+        case _: ConsensusViewRequest.type => 50
+        case _: ConsensusViewResponse     => 51
+        case _: ConsensusUserMsg          => 52
+      }, {
+        case 50 => viewRequestCodec.asInstanceOf[ByteCodec[ConsensusMsg]]
+        case 51 => viewResponseCodec.asInstanceOf[ByteCodec[ConsensusMsg]]
+        case 52 => userMsgCodec.asInstanceOf[ByteCodec[ConsensusMsg]]
+      }
+    )
+
+  private def runEventLoop(ref: Ref[List[NodeId]], self: NodeId) =
+    membership.events.foreach {
+      case MembershipEvent.Join(member) =>
+        ref.update(view => view :+ member.nodeId)
```

Comment:
This one's going to be inefficient, due to list's "performance" on appending. Alternatives:

- convert to prepend if you can live with reversed list
- use vector
- wait for chunk to support efficient append

Reply:
Sure, moved to Vector